File: pwiz/analysis/demux/SpectrumPeakExtractor.cpp
//
// $Id$
//
//
// Original author: Jarrett Egertson <jegertso .@. uw.edu>
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#define PWIZ_SOURCE

#include "SpectrumPeakExtractor.hpp"

namespace pwiz {
namespace analysis {
    SpectrumPeakExtractor::SpectrumPeakExtractor(const std::vector<double>& peakMzList, const pwiz::chemistry::MZTolerance& massError)
    {
        _numPeakBins = peakMzList.size();
        _ranges = boost::shared_array< std::pair<double, double> >(new std::pair<double, double>[_numPeakBins]);
        _maxDelta = 0.0;
        for (size_t i = 0; i < peakMzList.size(); ++i)
        {
            double peakMz = peakMzList[i];
            double deltaMz = peakMz - (peakMz - massError);
            if (deltaMz > _maxDelta) _maxDelta = deltaMz;
            _ranges[i].first = peakMz - deltaMz;
            _ranges[i].second = peakMz + deltaMz;
        }
        _minValue = _ranges[0].first;
        _maxValue = _ranges[_numPeakBins - 1].second;

        boost::shared_array< std::pair<double, double> > range_copy = boost::shared_array< std::pair<double, double> >(new std::pair<double, double>[_numPeakBins]);
        std::copy(_ranges.get(), _ranges.get() + _numPeakBins, range_copy.get());

        /* Verify non-overlapping of ranges. Overlap may be observed if the peaks are not centroided. Overlapping ranges are undesirable because they
         * could duplicate signal, which should be a conserved quantity. */
        for (size_t i = 0; i + 1 < _numPeakBins; ++i)
        {
            if (range_copy[i].second > range_copy[i + 1].first)
            {
                // find center and snap edges to this center
                double center = (range_copy[i].second + range_copy[i].first + range_copy[i + 1].second + range_copy[i + 1].first) / 4.0;
                _ranges[i].second = center;
                _ranges[i + 1].first = center;
                
            }
        }
    }

    void SpectrumPeakExtractor::operator()(msdata::Spectrum_const_ptr spectrum, MatrixType& m, size_t rowNum, double weight) const
    {
        // Puts intensities observed in the spectrum into bins specified by _ranges
        msdata::BinaryDataArray_const_ptr mzArray = spectrum->getMZArray();
        msdata::BinaryDataArray_const_ptr intensityArray = spectrum->getIntensityArray();

        // intitialize the filtered values to zero
        m.row(rowNum).setZero();

        size_t binStartIndex = 0;
        for (size_t queryIndex = 0; queryIndex < mzArray->data.size(); ++queryIndex)
        {
            // iterating through each "query" peak in the MS/MS spectrum to be filtered
            double query = mzArray->data[queryIndex];
            if (query < _minValue) continue;
            if (query > _maxValue) break;
            double minStart = query - _maxDelta;
            // move the starting point for this search and future searches to the first bin
            // that could possibly contain this peak
            for (; binStartIndex < _numPeakBins; ++binStartIndex)
            {
                if (_ranges[binStartIndex].first >= minStart) break;
            }
            // look forward for peak bins that contain this query
            for (size_t binIndex = binStartIndex; binIndex < _numPeakBins; ++binIndex)
            {
                // stop once past the query
                if (_ranges[binIndex].first > query) break;
                if (_ranges[binIndex].first <= query && query <= _ranges[binIndex].second)
                {
                    m.row(rowNum)[binIndex] += intensityArray->data[queryIndex];
                }
            }
        }
        // for (int i = 0; i < _numPeakBins; ++i) intensities[i] *= weight;
        m.row(rowNum) *= weight;
    }

    size_t SpectrumPeakExtractor::numPeaks() const
    {
        return _numPeakBins;
    }
} // namespace analysis
} // namespace pwiz

File: pwiz/analysis/Demux/SpectrumPeakExtractor.cs
using MathNet.Numerics.LinearAlgebra;
using Pwiz.Data.MsData.Spectra;
using Pwiz.Util.Chemistry;

namespace Pwiz.Analysis.Demux;

/// <summary>
/// Bins centroided peaks from a spectrum into a fixed grid of m/z windows. Port of cpp's
/// <c>pwiz/analysis/demux/SpectrumPeakExtractor.cpp</c>.
/// </summary>
/// <remarks>
/// Used by the demux algorithms to assemble the right-hand side <c>b</c> of the NNLS problem:
/// each row of the output matrix holds one input spectrum's intensities, summed into the
/// pre-defined per-peak bins. Bins straddle peaks within an <see cref="MZTolerance"/> window
/// (cpp's "delta = peakMz - (peakMz - massError)" idiom) and adjacent overlapping bins get
/// snapped to a common midpoint so a single peak can't double-count.
/// </remarks>
public sealed class SpectrumPeakExtractor
{
    private readonly (double Low, double High)[] _ranges;
    private readonly double _maxDelta;
    private readonly double _minValue;
    private readonly double _maxValue;

    /// <summary>Number of peak bins; equals the column count required for the output matrix.</summary>
    public int NumPeaks => _ranges.Length;

    /// <summary>Constructs an extractor with one bin per entry of <paramref name="peakMzList"/>.</summary>
    /// <param name="peakMzList">Sorted ascending peak m/z values; no duplicates.</param>
    /// <param name="massError">Tolerance window around each peak m/z.</param>
    public SpectrumPeakExtractor(IReadOnlyList<double> peakMzList, MZTolerance massError)
    {
        ArgumentNullException.ThrowIfNull(peakMzList);
        if (peakMzList.Count == 0)
            throw new ArgumentException("peakMzList must not be empty");

        int n = peakMzList.Count;
        _ranges = new (double, double)[n];
        double maxDelta = 0;
        for (int i = 0; i < n; i++)
        {
            double peakMz = peakMzList[i];
            // cpp: deltaMz = peakMz - (peakMz - massError) — the absolute Da equivalent of the
            // tolerance applied at this m/z (handles ppm vs. mz units transparently).
            double deltaMz = peakMz - SubtractTolerance(peakMz, massError);
            if (deltaMz > maxDelta) maxDelta = deltaMz;
            _ranges[i] = (peakMz - deltaMz, peakMz + deltaMz);
        }
        _maxDelta = maxDelta;
        _minValue = _ranges[0].Low;
        _maxValue = _ranges[n - 1].High;

        // Snap adjacent overlapping ranges to a shared midpoint so a single peak can't be
        // double-counted across two bins (cpp lines 47-57).
        for (int i = 0; i + 1 < n; i++)
        {
            if (_ranges[i].High > _ranges[i + 1].Low)
            {
                double center = (_ranges[i].High + _ranges[i].Low + _ranges[i + 1].High + _ranges[i + 1].Low) / 4.0;
                _ranges[i] = (_ranges[i].Low, center);
                _ranges[i + 1] = (center, _ranges[i + 1].High);
            }
        }
    }

    /// <summary>Bins <paramref name="spectrum"/>'s peaks into row <paramref name="rowNum"/> of
    /// <paramref name="matrix"/>, scaled by <paramref name="weight"/>. The row must have at
    /// least <see cref="NumPeaks"/> columns; existing values are overwritten.</summary>
    public void Extract(Spectrum spectrum, Matrix<double> matrix, int rowNum, double weight = 1.0)
    {
        ArgumentNullException.ThrowIfNull(spectrum);
        ArgumentNullException.ThrowIfNull(matrix);
        var mzArr = spectrum.GetMZArray();
        var intArr = spectrum.GetIntensityArray();
        if (mzArr is null || intArr is null) return;

        // Zero the destination row.
        for (int j = 0; j < matrix.ColumnCount; j++) matrix[rowNum, j] = 0;

        var mz = mzArr.Data;
        var inten = intArr.Data;
        int peakCount = System.Math.Min(mz.Count, inten.Count);

        int binStartIndex = 0;
        for (int q = 0; q < peakCount; q++)
        {
            double query = mz[q];
            if (query < _minValue) continue;
            if (query > _maxValue) break;

            double minStart = query - _maxDelta;
            // Advance binStartIndex past bins whose entire range is below the query window.
            for (; binStartIndex < _ranges.Length; binStartIndex++)
            {
                if (_ranges[binStartIndex].Low >= minStart) break;
            }
            // Add this peak's intensity to every bin whose [Low, High] contains it.
            for (int b = binStartIndex; b < _ranges.Length; b++)
            {
                if (_ranges[b].Low > query) break;
                if (_ranges[b].Low <= query && query <= _ranges[b].High)
                    matrix[rowNum, b] += inten[q];
            }
        }

        // Apply the row weight.
        for (int j = 0; j < NumPeaks; j++) matrix[rowNum, j] *= weight;
    }

    /// <summary>Computes <c>peakMz - massError</c> respecting the tolerance's units.</summary>
    private static double SubtractTolerance(double peakMz, MZTolerance tol) => tol.Units switch
    {
        MZToleranceUnits.Mz => peakMz - tol.Value,
        MZToleranceUnits.Ppm => peakMz - System.Math.Abs(peakMz) * tol.Value * 1e-6,
        _ => peakMz - tol.Value,
    };
}
